### PR TITLE
Support pre-ES-2015 browsers

### PR DESCRIPTION
**Pull request by tomek-he-him**
[Not all](http://caniuse.com/#search=promise) commonly-used browsers support promises natively. To get your lib to work in IE I have to install a global shim – which [isn’t a best practice](https://github.com/sindresorhus/object-assign/issues/10#issuecomment-65065859).

There are two reasonable options to deal with this:
- use a non-global shim like _[es6-promise](http://npm.im/es6-promise)_;
- use _[any-promise](http://npm.im/any-promise)_ to get any already-installed spec-compliant non-global polyfill.

I went with the second option.

### CODE REVIEW

**Comment by tomek-he-him** [comment]
:checkered_flag:

I’ve updated the readme as well.

**Comment by knpwrs** [comment]
I thought I had this documented somewhere, but I didn't want to include a `Promise` polyfill in order to keep this library as small as possible, with the understanding that somebody who understands promises would know that they have to supply their own polyfill if they wanted to use this library in older browsers.

I'll take a PR for a documentation change (or I'll otherwise update the docs at a later time), but I'd prefer to not include any more code as the point of this library is to be as small and simple as possible.

**Comment by tomek-he-him** [comment]
@KenPowers yeah, that’s why I used _any-promise_ which is also [very small](https://github.com/kevinbeaty/any-promise/blob/master/any-promise.js).

It’s not a polyfill – but just a library which returns any loaded CommonJS (non-global) polyfill. Global polyfills can be harmful as I posted in a link above (https://github.com/sindresorhus/object-assign/issues/10#issuecomment-65065859).

**Comment by tomek-he-him** [comment]
> I'll take a PR for a documentation change

You’ve already got it in the docs BTW: https://github.com/KenPowers/dr-promise/tree/e5e4e9be132bc9a742c3d877c474334d9ca81ac1#browser-support.

**Comment by knpwrs** [comment]
> You’ve already got it in the docs BTW

Ugh. You know it's early in the morning when I can't even properly read my own `README`.

I just took a look at `any-promise`. Interesting idea. I'll give this another look when I get home today.

**Comment by tomek-he-him** [comment]
:+1:

It’s 3 PM in Europe BTW. Sorry for waking you up :)

**Comment by knpwrs** [comment]
According to the [`any-promise`](https://github.com/kevinbeaty/any-promise/) [`README`](https://github.com/kevinbeaty/any-promise/blob/85b3b365fdde37117b65eb8a7c7daaebbca11a93/README.md) the browserify version always exports the global `Promise`. And indeed, the [`browser` field](https://github.com/kevinbeaty/any-promise/blob/85b3b365fdde37117b65eb8a7c7daaebbca11a93/package.json#L6) is set to [`any-promise-shim.js`](https://github.com/kevinbeaty/any-promise/blob/85b3b365fdde37117b65eb8a7c7daaebbca11a93/any-promise-shim.js) which just a single line:

``` js
module.exports = Promise;
```

Their `README` says to globally polyfill `Promise` as required for the case of using browserify.

Looking at [one of your repos](https://github.com/tomekwi/isomorphic-ensure) it looks like you're a webpack guy. Can you fill me in on the WebPack behavior in this case?

**Comment by tomek-he-him** [comment]
Wow! I didn’t read that carefully, I must tell you. I was in a hurry. But I did notice that my fork didn’t work out of the box in IE.

“Perhaps _any-promise_ needs a module that’s loaded – not just installed?” – so I thought. So I dropped a `require('es6-promise')` at the top of my tests. That worked – and I called it a day.

Today I gave that a closer look. It turns out that _browserify_ magically assigned `require('es6-promise').Promise` to a global variable `Promise`. No idea why – [only `require('es6-promise').polyfill()`](https://github.com/jakearchibald/es6-promise/tree/b49b2ad660d8dc676a8b01f01acbfab49e6b910b#auto-polyfill) should install a shim.

**By the way…**

I use _browserify_ for prototyping and running tests and _webpack_ for building real-world apps. I like the zero-config usage of _browserify_ – but there’s a lot of magic happening behind the scenes. Like the adventure I described above. I checked it out with _webpack_ – the global `Promise` was empty as I had expected. _webpack_ makes a tradeoff there – less magic, more config.

Having written [_isomorphic-ensure_](https://github.com/tomekwi/isomorphic-ensure) I had an opportunity to look what’s happening under the hood there. It seems like @kevinbeaty made `any-promise-shim.js` because [`any-promise.js:22`](https://github.com/kevinbeaty/any-promise/blob/85b3b365fdde37117b65eb8a7c7daaebbca11a93/any-promise.js#L22) is impossible to run through any bundler – be it _webpack_, _browserify_, _[nexe](https://github.com/jaredallard/nexe)_ or whatever else.

**Comment by tomek-he-him** [comment]
All in all, this PR doesn’t make sense anymore.

Perhaps this would make more sense?

``` js
require('dr-promise');  // Takes global `Promise`.

require('dr-promise/custom-promise')(
  require('es6-promise').Promise
);  // Takes custom one.
```

**Comment by tomek-he-him** [comment]
> I checked it out with webpack – the global Promise was empty as I had expected.

Sorry, I was wrong again. Webpack does the same. https://twitter.com/tomekwi_/status/646301704070275073.

**Comment by knpwrs** [comment]
There are two options I would suggest here.
1. Submit a pull request to `any-promise` to make it work with [`envify`](https://github.com/hughsk/envify), assuming there is something like [`envify`](https://github.com/hughsk/envify) for webpack.
2. Do nothing. That linked issue describing `Object.assign` applies less here. The `Promise` API is more or less final in that it is available unprefixed in enough browsers that if the spec were to change in a breaking way apps which rely on the spec as-is (with the expectation of running only in modern browsers) would break because of a native API.

**Comment by tomek-he-him** [comment]
> Submit a pull request to any-promise to make it work with envify

As far as I know, neither _webpack_ nor _browserify_ would swallow something like `require(process.env.MY_FAV_PROMISE)`. They need plain strings or something very close to that.

> The Promise API is more or less final

Yup. But many devs still use A+ promises like _q_ in their apps. AFAIK native Promises are not 100% A+ compliant.

I’m developing a library. While it’s sometimes fine to install a global polyfill in an app, a library shouldn’t do it.

**Comment by knpwrs** [comment]
Not necessarily `require(process.env.MY_FAV_PROMISE)`, but something like:

``` js
if (process.env.PROMISE_IMPL === 'bluebird') {
  module.exports = require('bluebird');
} else if (process.env.PROMISE_IMPL === 'rsvp') {
  module.exports = require('rsvp');
} // etc
```

It's more verbose up front, but with `envify` it would be reduced to a single `require` and export.

Another thought occurs, I'll have to think about it a bit more, but we could also potentially use `any-promise` as-is and suggest the use of [`aliasify`](https://github.com/benbria/aliasify). Tell the end user that if they want to use non-native promises that they should alias `any-promise` to their favorite promise library.

**Comment by tomek-he-him** [comment]
> we could also potentially use any-promise as-is and suggest the use of aliasify. Tell the end user that if they want to use non-native promises that they should alias any-promise to their favorite promise library.

That sounds very reasonable to me. 😊

**Comment by knpwrs** [comment]
I'll do some testing when I get home today.

**Comment by kevinbeaty** [comment]
> Submit a pull request to any-promise to make it work with envify, assuming there is something like envify for webpack.

I'd be open to this if it works.  I was not aware of `envify`.  The only purpose of the shim was just to get it to work at all in the browser.

**Comment by tomek-he-him** [comment]
> Not necessarily `require(process.env.MY_FAV_PROMISE)`, but something like:
> 
> ``` js
> if (process.env.PROMISE_IMPL === 'bluebird') {
>   module.exports = require('bluebird');
> } else if (process.env.PROMISE_IMPL === 'rsvp') {
>   module.exports = require('rsvp');
> } // etc
> ```

@kevinbeaty @KenPowers AFAIK, this would mean someone bundling the app would have to manually exclude _bluebird_, _rsvp_ and all other implementations he doesn’t use. Webpack certainly would require that.

I guess the _aliasify_ option would really be more convenient. Or just using [the `browser` field](https://github.com/substack/browserify-handbook/tree/e21ccd34f63be30baa387fa846874602d6150c08#browser-field) (there is an equivalent in webpack). That can be easily done today, with this PR merged.

Then all we’d really need is a doc update. @kevinbeaty if it works out I can mention the workflow in a PR to _any-promise_ too. 😃

---

BTW, there used to be a “😊” at the end – but I didn’t realize it’s a blush! 😀

**Comment by tomek-he-him** [comment]
> assuming there is something like envify for webpack.

There used to be an _[envify-loader](https://github.com/mjohnston/envify-loader)_ – but now there is a way to [use browserify transforms directly](https://github.com/webpack/transform-loader).

**Comment by kevinbeaty** [comment]
@tomekwi 

> Then all we’d really need is a doc update. @kevinbeaty if it works out I can mention the workflow in a PR to any-promise too. :blush:

Works for me.

**Comment by knpwrs** [comment]
> @kevinbeaty @KenPowers AFAIK, this would mean someone bundling the app would have to manually exclude bluebird, rsvp and all other implementations he doesn’t use. Webpack certainly would require that.

Just got done with my testing. At least with browserify, as long as `envify` and `uglifyify` are run as _global_ transforms (in that order), then you only need to install the promise implementation that you want to use. Additionally, setting `purge` on envify allows a default option.

My test was as follows:

First, install `envify`, `uglifyify`, and `bluebird`. Then, create a file `node_modules/muh-promise/index.js`:

``` js
if (process.env.PROMISE_IMPL === 'bluebird') {
  module.exports = require('bluebird');
} else if (process.env.PROMISE_IMPL === 'rsvp') {
  module.exports = require('rsvp');
} else {
  module.exports = global.Promise;
}
```

Then, create a file in the root directory `app.js`:

``` js
var Promise = require('muh-promise');
Promise.resolve(global.foo).then(function () {
  console.log('Foo!');
});
```

Now you can build this with global transforms and the `envify` purge option:

``` sh
PROMISE_IMPL=bluebird browserify -g [ envify purge ] -g uglifyify app.js
```

If `envify`'s `purge` option is not set, and there is no value for `PROMISE_IMPL`, then `browserify` will complain about not being able to find `rsvp` (since we only installed `bluebird`).

This configuration is also possible programmatically or through `package.json`:

``` js
b.transform(envify({
    _: 'purge'
}))
```

When `purge` is set but `PROMISE_IMPL` is not set then `muh-promise` will resolve to `global.Promise` (matching the current behavior of `any-promise`).

---

To recap:
1. Set `PROMISE_IMPL` and use `envify` and `uglifyify` to only require installation of a single promise library.
2. Set `envify`'s `purge` option to additionally allow the flexibility of not having to set `PROMISE_IMPL`, defaulting to `global.Promise`.

**Comment by tomek-he-him** [comment]
:+1:

**Comment by tomek-he-him** [comment]
:+1: for the _envify_ solution.

Yeah, I’ve tried out the _aliasify_ option – and it turns out that it doesn’t work as expected. _dr-promise_ is a dependency of my module _foo_. Browserify doesn’t let me `aliasify` _any-promise_ which is `require`d by _dr-promise_. I can only `aliasify` modules which I require directly from _foo_.

**Comment by knpwrs** [comment]
It should work if you run it as a global transform, which is what you want since you would want to alias `any-promise` everywhere.

**Comment by tomek-he-him** [comment]
@KenPowers cool, didn’t know about global transforms. (In webpack that’s the default and only behavior of loaders.)

But I understand [from _browserify’s_ docs](https://github.com/substack/node-browserify/tree/05d786175e13b873c49afc40170a0428eb56354c#btransformtr-opts) that I can use global transforms only programmatically and that I can’t  configure that in the `package.json` or somewhere. That doesn’t sound good. 😞

**Comment by knpwrs** [comment]
You can still use the `-g` option as I did in my `envify` + `uglifyify` example. So just install `browserify` as a dev dependency then run it as a script:

``` json
"scripts": {
  "build": "browserify -g aliasy"
},
"devDependencies": {
  "browserify": "..."
}
```

Then just run `npm run build`.

You can use the subargs syntax as I did for `envify` + `uglifyify`. I am not sure if Browserify will combine subargs / other args with `package.json` configuration or not.

**Comment by tomek-he-him** [comment]
:+1: I’ll try that out when I have a spare while. I’m quite in a hurry now.
